Add README instructions from prior documentation in Google Drive

chelseatroy opened the pull request:
### Goal(s) of this Pull Request:

- Move documentation that would still be relevant to an MLE end user who wants to use this template from the Google doc to the README for this repository.

### Example of how it works:

- Ideally, an MLE can follow the steps in this README and see both the metaflow flow and wandb dashboard successfully.

chelseatroy committed:
Add README instructions from prior documentation in Google Drive

chelseatroy committed:
whitespace

Dexterp37 commented on README.md:
```diff
@@ -1,2 +1,61 @@
-# moz-mlops
-A tooklit for productionizing machine learning models at Mozilla. 
+# Training and Deploying Models in The Cloud
+
+So you're a Mozillian with a machine learning model that you have futzed with locally, or maybe in Google Colab, or possibly even on your own cluster. You’re now ready for a more sophisticated model orchestration setup. 
+
+Excellent! You've found the toolset for this. We use deployed Metaflow flows with Outerbounds to do model orchestration, and we use Weights and Biases for experiment evaluation. The templates in this repository will help you integrate with both.
```
```suggestion

Mozilla uses Metaflow for model orchestration and Weights and Biases for experiment tracking.
The templates in this repository will help you integrate with both.
```

chelseatroy replied:
I don't think it makes it clear who the audience is, what this resource is for, or how much help it provides.

Dexterp37 replied:
We can assume the audience to be ML-practitioners at Mozilla: my main objective here is to make this readme drier/shorter. Feel free to reject my suggestion in favour of any other succinct head!

chelseatroy replied:
I did some cleaning up in the latest version!

Dexterp37 commented on README.md:
```diff
@@ -1,2 +1,61 @@
+
+Excellent! You've found the toolset for this. We use deployed Metaflow flows with Outerbounds to do model orchestration, and we use Weights and Biases for experiment evaluation. The templates in this repository will help you integrate with both.
+
+## But first...you need some stuff.
+
+### Most importantly, you need an account with Outerbounds (do not make this yourself).
+
+An admin from Mozilla’s MLOps team needs to make you a user account in our production Outerbounds instance. You can ask for help with this in the #mlops Slack channel.
+
+**The admins need to know two things:**
+1. If your data requires more restricted access than All-Mozilla. If so, they will handle getting a perimeter created for you, and will get back to you with your Outerbounds login information and perimeter name. 
+1. If you intend to run your flow on your own cluster or your own servers, rather than Outerbounds resources. If yes, they need to set that up. Let them know and they’ll work with you on that.
+
+Once MLOps comes back to you with a shiny new Outerbounds account and a perimeter for your team, you can sign in [here](https://docs.google.com/document/d/12S06Q-9xh6YGkyBpi6iGWGazYDivxl2fT4KaHtbrzV4/edit).
+
+Go to [this page right here](https://ui.desertowl.obp.outerbounds.com/dashboard/configure?location=local). You’ll see some instructions. You have already installed Outerbounds, so you get to skip that one. 
+Look for a dropdown that says “perimeter.” You want to make sure that, if your data has more restricted access than All-Mozilla, you have checked the perimeter name that you received from MLOps. If your data has all-Mozilla access, you can check “default” here.
+
+You should see an instruction to run a command line command that starts with `outerbounds configure`. Copy that whole thing and run it on your command line. What this does, is it makes a `config.json` file located at `~/.metaflowconfig` on your local machine. This is the default configuration for metaflow on your machine now, and unless told otherwise, metaflow flows on your machine will now alwa ...
```
I think this should live in [Confluence](https://mozilla-hub.atlassian.net/wiki/spaces/DATA/pages/522190969/MLOps), in its own page next to "Getting a W&B account". We can then change this to:

```suggestion
## Getting started

### Provisioning accounts
Please refer to [Mozilla Internal Documentation](LINK HERE) to have your accounts provisioned.
```

chelseatroy replied:
That makes sense to me, given that this repo is public and the steps here would be specific to Mozilla. Since I have _not_ set up that Confluence page yet and don't want it to block me on moving this over to the other repo, I'm going to table this suggestion for now and make the change _after_ I get this into the other repo.

chelseatroy replied:
In the future, @Dexterp37 I'd appreciate if you could state the reason for substantive changes. Not because I necessarily disagree, but because it transfers context. In this case I could tell, but in other cases, the context for why on changes like this stays in your head and all I see is 'make this change please.'

Dexterp37 replied:
Apologies, I believe my intent was clear because of my top comment [in the original doc](https://docs.google.com/document/d/12S06Q-9xh6YGkyBpi6iGWGazYDivxl2fT4KaHtbrzV4/edit#heading=h.ulsdk1khtl7i). Will make sure to be more explicit next time!

Dexterp37 commented on README.md:
```diff
@@ -1,2 +1,61 @@
+
+You should see an instruction to run a command line command that starts with `outerbounds configure`. Copy that whole thing and run it on your command line. What this does, is it makes a `config.json` file located at `~/.metaflowconfig` on your local machine. This is the default configuration for metaflow on your machine now, and unless told otherwise, metaflow flows on your machine will now alwa ...
+
+### Once you have done the above, you're ready to use the template in this repository.
+
+1. Install the requirements listed in the `requirements.txt` in the `templates` directory of this repo, and also add those to your project requirements files.
+2. Copy the `template_flow.py` file from ths repo into _your_ repo. From there, you can run `python template_flow.py run --offline True`.
+
+You will see your metaflow flow kick off with such proclamations as "The graph looks good!" and "Pylint is happy!", and then, [you can go here](https://ui.desertowl.obp.outerbounds.com/dashboard/runs) (to the “Runs” tab in the Outerbounds UI) to track the progress of the flow, see stderr and stdout, et cetera. 
+
+Eventually you should see the flow finish with "Task finished successfully." Once you see that, you know you've got a working minimal flow: you're now ready to start putting your model training code _into_ this flow to run it on the Outerbounds instance. Comments _inside_ the template file should help you understand where to put that code, and demonstrate some of the tools Metaflow makes availabl ...
```
```suggestion
### Running the template Metaflow flow

1. Install the requirements listed in [`templates/requirements.txt`](`templates/requirements.txt`) and also add those to your project requirements files.
2. Copy the `template_flow.py` file from this repo into _your_ repo. From there, you can run `python template_flow.py run --offline True`.

> [!TIP]
>The `--offline True` command line argument tells the template flow to not track experiments on W&B.

The Metaflow flow will run internal consistency checks and linting, and then provide a link to track the progress of the flow on our platform, see stderr and stdout, et cetera. 

Eventually you should see the flow finish with "Task finished successfully." Once you see that, you know you've got a working minimal flow: you're now ready to start putting your model training code _into_ this flow to run it remotely. Comments _inside_ the template file should help you understand where to put that code, and demonstrate some of the tools Metaflow makes available to you.
```

chelseatroy replied:
This one I'm gonna go ahead and commit before i move this over!

Dexterp37 commented on README.md:
```diff
@@ -1,2 +1,61 @@
+
+Eventually you should see the flow finish with "Task finished successfully." Once you see that, you know you've got a working minimal flow: you're now ready to start putting your model training code _into_ this flow to run it on the Outerbounds instance. Comments _inside_ the template file should help you understand where to put that code, and demonstrate some of the tools Metaflow makes availabl ...
+
+ ## Next: A Fancy Dashboard
```
```suggestion
### Adding experiment tracking
```

chelseatroy replied:
I am confident that "experiment tracking" means different things to different people. Within Mozilla, even, we have tools _outside_ the ML team (for example Nimbus) whose purpose would _also_ be described as this. I'm sensitive to the dangers of the use of overloaded terms to describe what's going on: in context, they can misdirect folks. I'm getting that you're not a fan of the colloquial tone I take in documentation (and I think that is a conversation rather than a series of commit suggestions), but I think we should opt for something more specific here regardless.

Dexterp37 replied:
I see your concern, "experiment" is indeed an overloaded term. However, experiment tracking is exactly the thing we're using W&B for. We not only produce a beautiful dashboard, but use it to make decisions on which experiment produced the best model given some metric. I don't think "A fancy dashboard" makes the cut, I'm open to options!

chelseatroy replied:
Fantastic. I came up with something else, reflected in the latest version!

Dexterp37 commented on README.md:
```diff
@@ -1,2 +1,61 @@
+
+ ## Next: A Fancy Dashboard
+
+An admin from Mozilla’s MLOps team needs to set you up with a Weights and Biases User account, and also a team if your team doesn’t have that yet, on Weights and Biases. Reach out to #mlops in Slack for help with this.
+
+MLOps needs to know:
+1. If you will need a service account for your Weights and Biases team—say, to set up CI or something. Seats cost significant amounts of money, so if you need extra accounts “just in case,” no you don’t. But if you are, for example, already set up with CI for your model and you want that on Weights and Biases, we can get you a service account for that.
```
```suggestion
```

This is already documented in Confluence. The initial provisioning accounts will take care of pointing users the right way.

chelseatroy replied:
In that case my move here would be to link to confluence instead, which, like the above, I'll execute after I move this over to the other repo. 

I _do_ want to mention in the docs _why_ some of the instructions live on Confluence, because it's not a great experience for people to have to go to multiple separate places to get the complete instructions. In this case because these are Mozilla-specific steps that live on Mozilla-specific confluence it makes total _sense_, but I'd want to _acknowledge_ the tradeoff.

Dexterp37 replied:
Good point. I'd try to keep this doc short though, adding a short note to the top in the provisioning accounts section

Dexterp37 commented on README.md:
```diff
@@ -1,2 +1,61 @@
+MLOps needs to know:
+1. If you will need a service account for your Weights and Biases team—say, to set up CI or something. Seats cost significant amounts of money, so if you need extra accounts “just in case,” no you don’t. But if you are, for example, already set up with CI for your model and you want that on Weights and Biases, we can get you a service account for that.
+
+Once you have the account, you're in luck; the `template_flow` already integrates with Weights and Biases for you. 
```
```suggestion
The `template_flow` already integrates with Weights and Biases for you. 
```

chelseatroy replied:
So this is an example of the sort of thing that I think is a conversation rather than a bunch of commit suggestions made without additional comment. 

It feels like you don't think there's a good reason I did it this way, other than maybe ignorance to convention, which feels like a lack of trust in my experience and my judgment. I would have expected this to be a question. "Hey, I noticed that the whole flow takes a colloquial tone. That's unconventional. Why did you make that choice?"

Dexterp37 replied:
> It feels like you don't think there's a good reason I did it this way,

I never said or implied that. Reviews are conversations and all the changes suggested here were made in the same spirit of the comments on the original document. The dropped text doesn't add much, _in my opinion of a non-native English speaker_, to the content: doesn't make me more able to reason about the flow and it contributes to the overall length of this document. I believe this is too long of a readme to explain how to use one template file.

chelseatroy replied:
So, the reason I did it this way is that I foresee a miscommunication with README users in which they think "The `template_flow` already integrates with Weights and Biases for you" means they don't need, or aren't blocked on, the account creation steps explained above. I think that we have _raised_ the risk of that miscommunication by moving the account creation text to Confluence, where they likely have not seen it by the time they get here. 

I believe taking a little extra time to mitigate risk of miscommunication is usually worth it. It has been my humble experience that I have lost _much_ more time to making up for incorrect assumptions made by documenters than I have to reading a _slightly_ longer piece ahead of using the code it describes.

Dexterp37 commented on README.md:
```diff
@@ -1,2 +1,61 @@
+
+Note the three environment variables listed at the top of the `train` step:
+
+- The WANDB_API_KEY you can get by going to this page, once your wandb credentials are set up (see step 1). Treat this with the sensitive that you would treat any API key. You will see in a later step how to use this string; for now, you can stick this in your local repository’s .env. 
```
```suggestion
- The `WANDB_API_KEY` you can get by going to [this page](https://wandb.ai/), once your wandb credentials are set up (see the account provisioning section). Treat this with the sensitive that you would treat any API key. You will see in a later step how to use this string; for now, you can stick this in your local repository’s .env. 
```

Dexterp37 commented on README.md:
```diff
@@ -1,2 +1,61 @@
+
+- The WANDB_API_KEY you can get by going to this page, once your wandb credentials are set up (see step 1). Treat this with the sensitive that you would treat any API key. You will see in a later step how to use this string; for now, you can stick this in your local repository’s .env. 
+
+- The WANDB_ENTITY is your wandb username (or, in the case of a service account, the service account username, which MLOps should have provided to you if you had them make a service account). 
```
```suggestion
- The `WANDB_ENTITY` is your wandb username or the service account username (provided to you by the MLOps upon request). 
```

chelseatroy replied:
Also adding this!

Dexterp37 commented on README.md:
```diff
@@ -1,2 +1,61 @@
+
+- The WANDB_ENTITY is your wandb username (or, in the case of a service account, the service account username, which MLOps should have provided to you if you had them make a service account). 
+
+- The WANDB_PROJECT is the name of the project that this run is for. You can see and create projects for your team at this URL: https://wandb.ai/YOUR_TEAM_NAME_HERE/projects.
```
```suggestion
- The `WANDB_PROJECT` is the name of the project that this run is for. You can see and create projects for your team at this URL: https://wandb.ai/YOUR_TEAM_NAME_HERE/projects.
```

Dexterp37 commented on README.md:
```diff
@@ -1,2 +1,61 @@
+You can customize what you send to Weights and Biases as well as the graphs and data that appears there, and the team maintains decent documentation on your customization options. The wandb engineering support team is also incredibly helpful. Talk to MLOps about your problem and they can get you added to our joint channel with wandb where folks are available to help.
+
+
+When it’s time to run your flow on outerbounds with your weights and biases dashboard set up, the Outerbounds instance will not have the three environment variables we just talked about. We are currently working on that with them for things like scheduled flows. In the meantime, when you kick off the flow from your local machine, you can specify the environment variables via our super-advanced te ...
```
```suggestion
To run the template flow and test the connection to W&B, specify the environment variables on the CLI:
```

chelseatroy replied:
I'd like to take this instance as an example of the kind of thing I'd would have liked to say in a conversation about why I write documentation this way for this project.

Engineers are not going to like that this is how they have to specify credentials, and they are right; it is objectively absurd, inconvenient, and furthermore insecure, as it means they'll have their unencrypted `WAND_API_KEY`s repeated ad infinitum in their console history. 

There should be, and will be, a better way to do this. _Acknowledging_ this kind of thing up front reassures our constituents that we know what this user experience is like, and we have plans to change it. That engenders trust in both our technical acumen and our empathy for their experience. Since it is our explicit goal to build close relationships without constituents as purveyors of an internal tool, that _branding_—"Hi! We're a team that's here to help you! We know this part fails a little bit in that mission, and we're working on it!" makes a difference in constituent experience at scale over time. 

It's not that I just don't know how documentation is usually written. I am, to my chagrin, aware.

Dexterp37 replied:
> Engineers are not going to like that this is how they have to specify credentials, and they are right; it is objectively absurd, inconvenient, and furthermore insecure, as it means they'll have their unencrypted `WAND_API_KEY`s repeated ad infinitum in their console history.

This is a problem that we'll need to solve and, once we do, we'll update the docs.
I don't feel strongly about how we pass API keys around in the example here, prepending them in the CLI is one way, but we could document other ways you'd recommend to pass them around instead!

> There should be, and will be, a better way to do this. _Acknowledging_ this kind of thing up front reassures our constituents that we know what this user experience is like, and we have plans to change it. That engenders trust in both our technical acumen and our empathy for their experience. Since it is our explicit goal to build close relationships without constituents as purveyors of an internal tool, that _branding_—"Hi! We're a team that's here to help you! We know this part fails a little bit in that mission, and we're working on it!" makes a difference in constituent experience at scale over time.

I believe the best way to acknowledge this problem is prioritize its resolution and address it. I don't think making this _specific document_ longer would help with building trust or empathy, but I'm happy to take a step back on this one.

chelseatroy replied:
> I don't feel strongly about how we pass API keys around in the example here, prepending them in the CLI is one way, but we could document other ways you'd recommend to pass them around instead!

I don't disagree with the decision, in context, to prepend the key to the command: it is the option we have in the absence of Outerbounds having provided the alternative yet, which is why you made the (in my opinion pragmatic) decision to do it this way, and why we're recommending it at the moment. The decision to do this is not what's inconvenient and absurd; it's the context _forcing_ us to do it that's inconvenient and absurd. I apologize for not making this clearer. 

> I believe the best way to acknowledge this problem is prioritize its resolution and address it. I don't think making this specific document longer would help with building trust or empathy, but I'm happy to take a step back on this one.

I disagree. First, in this specific case, we don't have control over the resolution: _our_ implementation isn't the blocker. _Outerbounds's_ is. But because we don't want to sour people on the tool, I agree with you that we shouldn't mention this to constituents. But I think we can still acknowledge "this is bad, we know, we're working on it." 

Second, I've seen a lot of cases where leadership teams—which I think describes any team that has power over a group of other people, so I think it includes library maintainers—lost the confidence of their constituents because, despite the fact that they were working hard to fix something, they didn't tell their constituents that. To constituents, that's indistinguishable from being ignored, which is what they'll often assume without (and sometimes unfortunately even _with_) a pre-established trust relationship. 

There'll be times in the future when we, as leaders of this project, for legal or regulatory or privacy or security reasons, _have_ to keep our work confidential and just...absorb ire from our constituents over the status quo, or let them distrust that we plan to change it. But I'd like to minimize those occasions, and not do it any more often than we have to. And when it happens, a _track record_ of proactive transparency will buy us time.

Dexterp37 commented on README.md:
```diff
@@ -1,2 +1,61 @@
+(note that we have changed `offline` here to false—that means we _do_ want our flow to integrate with Weights and Biases!)
+
+
+
```
```suggestion
```

Dexterp37 commented on README.md:
````diff
@@ -1,2 +1,61 @@
+```bash
+WANDB_API_KEY=your-key-here WANDB_ENTITY=ctroy WANDB_PROJECT=mlops-codecopilot-demo python your-flow.py run --offline False
+```
+(note that we have changed `offline` here to false—that means we _do_ want our flow to integrate with Weights and Biases!)
````
```suggestion
> [!NOTE]  
> We have changed `offline` here to false: hat means we _do_ want our flow to integrate with Weights and Biases!
```

chelseatroy committed:
Update README.md to add TIP formatting

Co-authored-by: Alessio Placitelli <a.placitelli@a2p.it>

chelseatroy committed:
Apply suggestions from code review for clarity and formatting

Co-authored-by: Alessio Placitelli <a.placitelli@a2p.it>

chelseatroy commented:
@Dexterp37 given the volume of conversation we have going on here, together with the fact that I am moving this file over to the other repo, what I'm going to do is: 

1. Start a PR for adding this README in the other repo.
2. Link back, in the PR description for that PR, to this PR. 
3. Keep changing _this_ version of the document and continue the conversation in _this_ PR.
4. When this PR reaches resolution/approval, copy the latest version of the document from _this_ repo to _that_ repo.
5. Then I will [archive](https://docs.github.com/en/repositories/archiving-a-github-repository/archiving-repositories) this repository to communicate clearly that it is no longer actively maintained, but the PR conversations will still be visible to someone who clicks the link in the new repo.

chelseatroy committed:
Two updates:

+ More descriptive titles
+ Terse introductions
+ Moving Mozilla-specific instructions into the confluence

Dexterp37 reviewed (approved):
Let's move this over to the other repo and keep refining there.

chelseatroy commented:
OK! It's been moved!

chelseatroy commented:
Closing this PR, as we are moving the document to the [mozmlops](https://github.com/mozilla/mozmlops) repo.
